src/zap/zapinfo.cpp

```diff
@@ -1915,20 +1928,9 @@ void ZapInfo::getFunctionFixedEntryPoint(CORINFO_METHOD_HANDLE   ftn,
 
     // We can only embed entrypoints from the module being NGened since we do not support mapping of external 
     // import thunks to MethodDesc. It should be ok since the delegate targets are typically from the same module.
-    void * entryPointToEmbed = embedDirectCall(ftn, CORINFO_ACCESS_ANY, FALSE);
-
-    if (entryPointToEmbed != NULL)
-    {
-        pResult->accessType   = IAT_VALUE;
-        pResult->addr         = entryPointToEmbed;
-    }
-    else
+    if (!embedDirectCall(ftn, CORINFO_ACCESS_ANY, FALSE, pResult))
     {
-        ZapImport * pImport = m_pImage->GetImportTable()->GetFunctionEntryImport(ftn);
-        AppendConditionalImport(pImport);
-        
-        pResult->accessType   = IAT_PVALUE;
-        pResult->addr         = pImport;
+        embedFunctionEntryImport(ftn, pResult);
```

Comment:
There is similar problem on non-x86 even for regular calls. The JIT does tail-call optimization for regular calls on non-x86: If you have code like:

```
static void foo()
{
    ... stuff ... 
    bar();
}
```

The JIT will generate code like this:

```
... stuff ...
jmp [bar]
```

You won't be able to disassemble the callsite in this case ... the address needs to be passed in explicitly. Once you hit it on x64, you can disable the tailcall optimizations to workaround it (canTailCall - it is what R2R does) ... but I think it will need to be fixed.

Reply:
I don't fully understand what the problem is I think. I think I understand the point about the tail calls, but not how it relates to the code you marked for this comment. That code was there before, I just deduplicated it between the two "get entrypoint" methods.
About the tail call - we do have logic elsewhere to figure out jump/call targets - we basically look at the instruction and decide which one to use. Assuming JIT always uses the same type of call/jump, we should still be able to do that... right?
If not then we can either:
- disable the tails calls as you mention. I'm wondering if doing it globally for everything like R2R does is not too much of an overkill. For this change we could scope it to external methods only...
- Introduce different thunk, which would only server as a way to identify the callee. We would still do indirect call, but starting with it pointing to the thunk, the worker would use that to figure out what is being called and then patch the indirection cell instead (it should be possible to make it such that by knowing the thunk we can figure out the matching indirection for it). 

Not sure which one is better. Disabling tails calls is definitely less complex... but the perf hit might be "interesting enough".

Reply:
> how it relates to the code you marked for this comment

Sorry, the comment should have been at different spot.

> JIT always uses the same type of call/jump, we should still be able to do that... right?

The problem is that you do not get the right callsite to disassembly when the JIT jumps to the method instead of calling it.

> but the perf hit might be "interesting enough".

Yes...

Reply:
So what I'm getting is that you would prefer the "thunk" solution. Which there are 2 flavors of:
1) Callsite uses indirection cell, but that points to thunk which contains the additional information (somehow), but when we resolve it we go and patch the indirection cell leaving the thunk as is - so the thunk becomes RX only.
2) Callsite uses direct call to a thunk, which inside uses indirect call to the stub. When comes to patching we still patch the indirection cell, but this whole thing is hidden from the callsite (as in the callsite itself didn't need to change from what it is today). The thunk is again RX only.

Reply:
There is also 3rd option that the JIT will use a agreed upon contract for tailcalls - always pass the indirection in given register. The JITed code for tailcalls would be:
```
...
lea r10, [indirection]
jmp [r10]
```
Either option would work. Your option (1) has best steady state so it is what I would lean towards.